File: Cpp/103_Binary_Tree_Zigzag_Level_Order_Traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        vector<vector<int> > result;
        if(!root) return result;
        
        queue<TreeNode*> q;
        q.push(root);
        bool is_forward_direction = true;
        while(!q.empty()){
            int n = q.size();
            vector<int> temp(n, 0);
            for(int i = 0; i<n; ++i){
                int index = is_forward_direction ? i : n-i-1;
                TreeNode* current = q.front();
                temp[index] = current->val;
                q.pop();
                if(current->left) q.push(current->left);
                if(current->right) q.push(current->right);
            }
            result.push_back(temp);
            is_forward_direction ^= true;
        }
        return result;
    }
};

File: Kotlin/103_Binary_Tree_Zigzag_Level_Order_Traversal.kt
/**
 * Example:
 * var ti = TreeNode(5)
 * var v = ti.`val`
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */
class Solution {
    fun zigzagLevelOrder(root: TreeNode?): List<List<Int>> {
        val result = ArrayList<List<Int>>()
        if(root == null) return result
    
        val q = ArrayDeque<TreeNode>()
        var isForward = true
        q.offer(root)
        
        while(!q.isEmpty()){
            val n = q.size
            val temp = ArrayList<Int>()
            for(i in 0 until n){
                val index = if(isForward) i else n - i - 1
                val current = q.poll()
                temp.add(current.`val`)
                current.left?.let{q.offer(current.left)}
                current.right?.let{q.offer(current.right)}
            }
            
            if(isForward) result.add(temp)
            else result.add(temp.reversed())
            isForward = !isForward
        }

        return result
    }
}